### PR TITLE
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

### PR DESCRIPTION
**Pull request by shamser**
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

```
Set tautological-compare compiler flag to be an error so 'this==null' comparision is treated as errors
tautlogical-compare also treats &reference==null as errors also
Code to remove &reference==null check by either removing the check where it is unnecessary or converting the reference to a pointer
Remove unnecessary check of IPropertyTree & for nullness that was returned in IPropertyTreeIterator.
ForEach() calls IPropertyTreeIterator's isValid member function to ensure item is vaid.
IAtom: Remove multiple aliases to access getNamePtr() from IAtom
Add global str(IAtom *) function to safely access getNamePtr() even if IAtom * argument is null
IIdAtom: Remove multiple aliases to access getNamePtr() from IAtom
Add global str(IIdAtom *) function to safely access getNamePtr() even if IIdAtom * argument is null
Replace all usages of the aliases removed previously with a call to the newly create str() function
IIdAtom: remove the lower() member variable and replace with global lower(IIdAtom *p) function
Modify cmake commonSetup.cmake to have this==null treated as compilation errors by CLang compilers
Remove checks of reference variable’s address
secuser pointer checked for nullness before passing in argument as reference (working.cpp: CSecureConstWUIterator::checkScope)
user pointer checked for nullness before passing in argument as reference (ws_DfuService.cpp: getFilePermissions)
Change user to reference in getFilePermission() (ws_dfuService.cpp). The caller checks user for nullness, so appropriate to have caller pass user as a reference.
modify declaration of str() and lower() in jhash.cpp with jlib_decl (to expose externally in dlls)
```

### CODE REVIEW

**Comment by shamser** [comment]
Please see original code review comments here: https://github.com/hpcc-systems/HPCC-Platform/pull/7410.  @richardkchapman

File: ecl/hql/hqlparse.cpp [hunk 1]
```diff
@@ -6,7 +6,7 @@
     you may not use this file except in compliance with the License.
     You may obtain a copy of the License at
 
-       http://www.apache.org/licenses/LICENSE-2.0
+       http://www.apache.org/licenses/LICENSE-2.04
```
Comment (richardkchapman) [inline ecl/hql/hqlparse.cpp]: This change looks unrelated

**Comment by richardkchapman** [comment]
A general comment: you have squashed down most of the original commits to a single one - this is often a good idea but in a case like this where there are two distinct phases to the change (the references that were done manually/carefully/each on their own merits, and need careful review, versus the str() change that was done automatically and does not need such careful review), it's a lot easier to review (and indeed to refer back to a change in the history should we need to) if you keep them separate.

Not worth fixing now since the squash has been done, but something to bear in mind for the future.

I can merge as it is, though there are a couple of minor comments you may want to address first. It's going to need a manual merge as it stands, since there is a clashing change merged into master since you branched/last rebased. I can do that (the clash is trivial) but normally I ask people to rebase and repush in such cases.

File: ecl/hql/hqlexpr.ipp [hunk 2]
```diff
@@ -399,7 +399,7 @@ protected:
 
 public:
     virtual IHqlExpression *clone(HqlExprArray &newkids);
-    virtual IAtom * queryName() const { return id->lower(); }
+    virtual IAtom * queryName() const { return id ? id->queryLower() : NULL; }
```
Comment (richardkchapman) [inline ecl/hql/hqlexpr.ipp]: I'm curious why you have changed some instances of lower(id) to the construct "id ? id->queryLower() : NULL" and others to say "lower(id)"

Would be better to use lower(id) in all cases I think.

File: ecl/hql/hqlparse.cpp (hunk 1 above)
Comment (shamser) [inline ecl/hql/hqlparse.cpp]: It is unrelated.  Fixed.

**Comment by shamser** [comment]
@richardkchapman Noted regarding keeping unrelated commits separate.  Code amended as per code review comments.

File: system/jlib/jhash.hpp [hunk 3]
```diff
@@ -49,11 +49,8 @@ interface jlib_decl IAtom : extends IMapping
 {
  public:
     virtual const char * getNamePtr() const = 0;
```
Comment (ghalliday) [inline system/jlib/jhash.hpp]: can we rename this function to something that would be sensible to call.

e.g., queryStr()

Some of the instances that have been converted could call this directly if we know that this != NULL.

File: common/thorhelper/thorrparse.cpp [hunk 4]
```diff
@@ -2022,7 +2022,7 @@ void RegexNamedPattern::toXMLattr(StringBuffer & out, RegexXmlState & state)
 {
     RegexPattern::toXMLattr(out, state);
     if (def->queryName())
-        out.appendf(" name=\"%s\"", def->queryName()->str());
+        out.appendf(" name=\"%s\"", str(def->queryName()));
```
Comment (ghalliday) [inline common/thorhelper/thorrparse.cpp]: could call def->queryName()->queryStr() since cannot be null

File: ecl/hql/hqlatoms.hpp [hunk 5]
```diff
@@ -448,7 +448,7 @@ extern HQL_API IAtom * xmlnsAtom;
 extern HQL_API IAtom * _xmlParse_Atom;
 extern HQL_API IAtom * xpathAtom;
 
-inline bool isInternalAttributeName(IAtom * name) { return (name->str()[0] == '$'); }
+inline bool isInternalAttributeName(IAtom * name) { return (str(name)[0] == '$'); }
```
Comment (ghalliday) [inline ecl/hql/hqlatoms.hpp]: minor: could call ->queryStr() since it would crash if it was NULL

File: ecl/hql/hqlerror.cpp [hunk 6]
```diff
@@ -206,7 +206,7 @@ class HQL_API FileErrorReceiver : public ErrorReceiverSink
 
         StringBuffer msg;
         error->errorMessage(msg);
-        if (!filename) filename = isError(severity) ? "" : *unknownAtom;
+        if (!filename) filename = isError(severity) ? "" : str(unknownAtom);
```
Comment (ghalliday) [inline ecl/hql/hqlerror.cpp]: unknownAtom->queryStr()

Comment (ghalliday) [inline ecl/hql/hqlerror.cpp]: I won't point any more out unless they are really critical.  Not really necessary to change any of them.

File: ecl/hql/hqlexpr.cpp [hunk 7]
```diff
@@ -7944,10 +7944,10 @@ void CHqlScope::throwRecursiveError(IIdAtom * searchName)
     StringBuffer filename;
     if (fullName)
         filename.append(fullName).append('.');
-    filename.append(*searchName);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: This change looks suspicious.  It should be 
   filename.append(str(searchName));
or even
   filename.append(searchName)

Previously the operator * was defined so that *myatom was the same as myatom->str();

*str(myAtom) will return just the 1st character

File: ecl/hql/hqlexpr.cpp [hunk 8]
```diff
@@ -7944,10 +7944,10 @@ void CHqlScope::throwRecursiveError(IIdAtom * searchName)
...
+    filename.append(*str(searchName));
 
     StringBuffer msg("Definition of ");
-    msg.append(*searchName).append(" contains a recursive dependency");
+    msg.append(*str(searchName)).append(" contains a recursive dependency");
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: same here

File: ecl/hql/hqlexpr.cpp [hunk 9]
```diff
@@ -8136,13 +8136,13 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
     StringBuffer filename;
     if (fullName)
         filename.append(fullName).append('.');
-    filename.append(*searchName);
+    filename.append(*str(searchName));
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: and here, and several lines below.

File: ecl/hql/hqlexpr.ipp [hunk 10]
```diff
@@ -1630,7 +1630,7 @@ public:
     virtual IHqlExpression *addOperand(IHqlExpression *field);
     virtual IHqlExpression *clone(HqlExprArray &newkids);
     virtual bool equals(const IHqlExpression & other) const;
-    virtual IAtom * queryName() const { return unnamedId->lower(); }
+virtual IAtom * queryName() const { return unnamedId ? unnamedId->queryLower() : NULL; }
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.ipp]: minor: indentation

File: esp/platform/sechandler.cpp [hunk 11]
```diff
@@ -95,6 +95,8 @@ bool SecHandler::authorizeSecFeature(const char * pszFeatureUrl, const char* Use
     if(pSecondaryUser.get()== NULL)
     {
         pSecondaryUser.setown(m_secmgr->createUser(UserID));
+	if (!pSecondaryUser.get())
```
Comment (ghalliday) [inline esp/platform/sechandler.cpp]: minor: indentation (this line and next)
trivial: you do not need to call .get()

File: esp/services/ws_access/ws_accessService.cpp [hunk 12]
```diff
@@ -727,7 +727,8 @@ bool Cws_accessEx::onAddUser(IEspContext &context, IEspAddUserRequest &req, IEsp
             cred.setPassword(pass1);
         try
         {
-            secmgr->addUser(*user.get());
+	    if (user.get())
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp]: minor: indentation

**Comment by ghalliday** [comment]
@shamser good work.  I've had an initial scan through and picked out a few issues - only one serious.
I'll have another more careful review once those are addressed.

File: ecl/hql/hqlgram.y [hunk 13]
```diff
@@ -3931,22 +3931,22 @@ attrib
     : knownOrUnknownId EQ UNKNOWN_ID        
                         {
                             parser->reportWarning(CategoryDeprecated, SeverityError, WRN_OBSOLETED_SYNTAX,$1.pos,"Syntax obsoleted; use alternative: id = '<string constant>'");
-                            $$.setExpr(createAttribute($1.getId()->lower(), createConstant(*$3.getId())));
+                            $$.setExpr(createAttribute(lower($1.getId()), createConstant(*str($3.getId()))));
```
Comment (ghalliday) [inline ecl/hql/hqlgram.y]: createConstant(str(...)) no *

**Comment by richardkchapman** [comment]
Looks like you changed once instance of the "id ? id->queryLower() : NULL; " paradigm - but there were several more.

**Comment by shamser** [comment]
@richardkchapman @ghalliday Thanks for your comments.  Changes as per code review comments complete.

File: ecl/hql/hqlatoms.hpp [hunk 14]
```diff
@@ -448,7 +448,7 @@ extern HQL_API IAtom * xmlnsAtom;
 extern HQL_API IAtom * _xmlParse_Atom;
 extern HQL_API IAtom * xpathAtom;
 
-inline bool isInternalAttributeName(IAtom * name) { return (name->str()[0] == '$'); }
+inline bool isInternalAttributeName(IAtom * name) { return name ? false : (name->queryStr()[0] == '$'); }
```
Comment (ghalliday) [inline ecl/hql/hqlatoms.hpp]: trivial: name cannot be NULL, so code can be simplified

File: ecl/hql/hqlerror.cpp [hunk 15]
```diff
@@ -206,7 +206,7 @@ class HQL_API FileErrorReceiver : public ErrorReceiverSink
 
         StringBuffer msg;
         error->errorMessage(msg);
-        if (!filename) filename = isError(severity) ? "" : *unknownAtom;
+        if (!filename) filename = isError(severity) || unknownAtom ? "" : unknownAtom->queryStr();
```
Comment (ghalliday) [inline ecl/hql/hqlerror.cpp]: The test for unknownAtom is unnecessary, and the wrong way around.

File: ecl/hql/hqlatoms.hpp (hunk 14 above)
Comment (ghalliday) [inline ecl/hql/hqlatoms.hpp]: (test is also the wrong way round)

**Comment by ghalliday** [comment]
@shamser I checked more carefully this time.  Just two issues picked up.

There will also need to be a 5.4 pull request which just adds the str() and lower() functions to avoid future merge errors.

We are also likely to need a pull request to the legacy repository (ln) to fix any instances there.

**Comment by richardkchapman** [comment]
I'm getting compilation errors with this branch:

/Users/rchapman/git/HPCC-Platform/plugins/mysql/mysqlembed.cpp:321:51: error: no member named 'str' in 'IAtom'
        msg.appendf(" for field %s", field->name->str());
/Users/rchapman/git/HPCC-Platform/plugins/mysql/mysqlembed.cpp:607:60: error: no member named 'str' in 'IAtom'
            failx("Error fetching column %s", field->name->str());

I'm guessing you compiled on a system with mysql not installed, so the optional MySQL plugin will have been skipped. You should check for similar issues in the other embedded language plugins.

**Comment by shamser** [comment]
@richardkchapman plugins checked to make sure correct member function called.  Just found the 2 you discovered.

**Comment by richardkchapman** [comment]
Merged manually to address required rebase
